upd/utils.py

```diff
@@ -0,0 +1,28 @@
+def convert_indices(N, *args):
+    """
+    Convertin indices from 4d array to 2d numpy array and vice-versa
+    :param N: size of corresponding _4d_ matrix: int
+    :param args: indices i, j, k, l or p,q: int
+    :return: list of converted indices: list
+    """
+    for elem in args:
+        if not isinstance(elem, int):
+            raise TypeError("Wrong indices")
+        if elem >= N:
+            raise TypeError("index is greater than size of the matrix")
```

Comment:
For two-particle sparse arrays the elements can be less than N*N. I would propose that this check goes after "elif len(args) == 2:", and another before "if len(args) == 4:" using "if elem >= N*N:" for the two-particle array indices.